narwhal/executor/src/subscriber.rs

```diff
@@ -243,151 +275,311 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
                 .subscriber_certificate_latency
                 .observe(cert.metadata.created_at.elapsed().as_secs_f64());
 
-            for (digest, (worker_id, _)) in cert.header.payload.iter() {
+            for (digest, (_, _)) in cert.header.payload.iter() {
                 self.metrics.subscriber_processed_batches.inc();
-
-                let mut workers = self.network.workers_for_certificate(cert, worker_id);
-
-                workers.shuffle(&mut ThreadRng::default());
+                let batch = fetched_batches
+                    .get(digest)
+                    .expect("[Protocol violation] Batch not found in fetched batches from certificate authors");
```

Comment:
Is the comment inside `expect()` accurate? I think the failure here actually means `fetch_batches_from_worker()` does not return all requested batches.